src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -29,45 +33,135 @@
+			assertTrue("rendezvous connection should be open", rendezvousConnection != null && rendezvousConnection.isOpen());
+			clientConnection.closeAsync().join();
+			
+			assertTrue("listener should still be open", listener.isOnline());
+			assertFalse("client connection should be closed", clientConnection.isOpen());
+			return rendezvousConnection;
+		}).thenCompose(rendezvousConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("rendezvous connection should be closed implicitly when client connection closes", rendezvousConnection.isOpen());
+			});
+		});
+		
+		// Rendezvous initiates closing
+		CompletableFuture<Void> rendezvousClosing = client.createConnectionAsync().thenCombineAsync(listener.acceptConnectionAsync(), (rendezvousConnection, clientConnection) -> {
+			assertTrue("client connection should be open", clientConnection != null && clientConnection.isOpen());
+			assertTrue("rendezvous connection should be open", rendezvousConnection != null && rendezvousConnection.isOpen());
+			rendezvousConnection.closeAsync().join();
+			
+			assertTrue("listener should still be open", listener.isOnline());
+			assertFalse("rendezvous connection should be closed", rendezvousConnection.isOpen());
+			return clientConnection;
+		}).thenCompose(clientConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("client connection should be closed implicitly when rendezvous connection closes", clientConnection.isOpen());
+			});
+		});
+		
+		CompletableFuture.allOf(clientClosing, rendezvousClosing).whenComplete(($void, ex) -> {
+			if (ex != null) {
+				fail(ex.getMessage());
+			}	
+			listener.close();
+		}).join();
 	}
 	
 	@Test
-	public void acceptWebSocketConnectionTest() {
-		CompletableFuture<Boolean> checkSocketConnectionTask = new CompletableFuture<Boolean>();
-		CompletableFuture<HybridConnectionChannel> conn = listener.acceptConnectionAsync();
-		CompletableFuture<HybridConnectionChannel> clientConnectionTask = client.createConnectionAsync();
-		conn.thenAccept((connection) -> {
-			checkSocketConnectionTask.complete(true);
-			clientConnectionTask.thenAccept(clientConnection -> clientConnection.closeAsync());
+	public void acceptHandlerTest() throws InterruptedException, ExecutionException, TimeoutException {
+		HybridConnectionListener listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+		AtomicInteger handlerExecuted = new AtomicInteger(0);
+		listener.openAsync().join();
+		
+		// Handler accepts the connection
+		listener.setAcceptHandler(context -> {
+			handlerExecuted.incrementAndGet();
+			return true;
 		});
-		assertTrue("Listener failed to accept connections from sender in webSocket mode.", checkSocketConnectionTask.join());
+		
+		client.createConnectionAsync().whenComplete((clientConnection, ex) -> {
```

Comment:
>whenComplete [](start = 33, length = 12)

who waits on the CF<Void> that represents this createConnectionAsync+whenComplete? If nobody ever waits on it would a failedAssert still cause the test to fail? (I don't think it would, personally)